### PR TITLE
issue: 4245 : A guest should not be able to add kudos to a restricted work

### PR DESCRIPTION
**Pull request by zz9pzza**
https://code.google.com/p/otwarchive/issues/detail?id=4245

### CODE REVIEW

File: app/controllers/kudos_controller.rb [hunk 1]
```diff
@@ -36,15 +36,18 @@ def create
...
-          elsif @kudo.creator_of_work?
+            msg = 'You have already left kudos here. :)'
+          end
+          if @kudo.creator_of_work?
             msg = "You can't leave kudos on your own work."
-          else
-            msg = "We couldn't save your kudos, sorry!"
+          end
+          if !current_user.present? && commentable.restricted?
+            msg = "You can't leave guest kudos on a restricted work"
```
Comment (scottsds) [inline app/controllers/kudos_controller.rb]: I know it's a tiny detail, but could you put a period on these messages, so they match some of the other warnings?

File: app/models/kudo.rb [hunk 2]
```diff
@@ -42,6 +43,15 @@ def cannot_be_author
     end
   end
 
+  def guest_cant_restricted
+    commentable = commentable_type.classify.constantize.
+                  find_by_id(commentable_id)
+    if pseud.nil? && commentable.restricted?
+      errors.add(:guest_on_restricted,
+                 ts("^You can't leave guest kudos on a restricted work"))
```
Comment (scottsds) [inline app/models/kudo.rb]: Period on this as well.

File: public/javascripts/application.js [hunk 3]
```diff
@@ -437,6 +437,9 @@ $j(document).ready(function() {
         if (data.errors && data.errors.cannot_be_author) {
           msg = "You can't leave kudos on your own work.";
         }
+        if (data.errors && data.errors.guest_on_restricted) {
+          msg = "You can't leave guest kudos on a restricted work";
```
Comment (scottsds) [inline public/javascripts/application.js]: And on this too, please!

**Comment by scottsds** [comment]
Just a general comment, perhaps it might be better to rename the 'msg' variable to something more clear? Like 'warning_messages' or 'error_messages' (since they're displayed as an error?). Even if you don't want to rename the variable, it needs to be pluralized since it can contain multiple messages.

**Comment by zz9pzza** [comment]
Scott it can only hold a single message.

File: app/models/kudo.rb [hunk 4]
```diff
@@ -3,6 +3,7 @@ class Kudo < ActiveRecord::Base
   belongs_to :commentable, :polymorphic => true
 
   validate :cannot_be_author
+  validate :guest_cant_restricted
```
Comment (sarken) [inline app/models/kudo.rb]: As long as you're editing, this method name is kind of opaque. Something like `guest_cannot_kudos_restricted_work` would probably be better -- it's longer, but it's very clear about what it's checking, and the examples in the [custom methods section of the RoR guide](http://guides.rubyonrails.org/active_record_validations.html#custom-methods) don't shy away from long names, either.

**Pull request merged**